src/traces/treemap/style.js

```diff
@@ -35,39 +35,65 @@ function styleOne(s, pt, trace, opts) {
-        var step = trace.marker.opacitystep;
-
-        return n === 0 ? base :
-            Math.max(0, Math.min(1, n * step));
-    };
+    var fillColor = cdi.color;
 
     if(hovered) {
         lineColor = trace._hovered.marker.line.color;
         lineWidth = trace._hovered.marker.line.width;
-        opacity = trace._hovered.marker.opacity;
     } else {
-        if(helpers.isHierarchyRoot(pt)) {
+        var isRoot = helpers.isHierarchyRoot(pt);
+
+        if(!trace._hasColorscale) {
+            if(pt.onPathbar) {
+                // Combining colors with the background color.
+                // Better not to have transparent segments.
+                // Also this helps pathbar texts appear nicely below previous segments.
+                fillColor = Color.combine(Color.addOpacity(trace._backgroundColor, 1 - trace.pathbar.opacity), fillColor);
+            } else {
+                var depthfade = trace.marker.depthfade;
+                if(depthfade) {
+                    var fadedColor = Color.combine(Color.addOpacity(trace._backgroundColor, 0.75), fillColor);
+                    var n;
+
+                    if(depthfade === true) {
+                        var maxDepth = helpers.getMaxDepth(trace);
+                        if(isFinite(maxDepth)) {
+                            if(helpers.isLeaf(pt)) {
+                                n = 0;
+                            } else {
+                                n = (trace._maxVisibleLayers) - (pt.data.depth - trace._entryDepth);
+                            }
+                        } else {
+                            n = pt.data.height + 1;
+                        }
+                    } else { // i.e. case of depthfade === 'reversed'
+                        n = pt.data.depth - trace._entryDepth;
+                        if(!trace._atRootLevel) n++;
+                    }
+
+                    if(n > 0) {
+                        for(var i = 0; i < n; i++) {
+                            var ratio = 0.5 * i / n;
+                            fillColor = Color.combine(Color.addOpacity(fadedColor, ratio), fillColor);
+                        }
+                    }
+                }
+            }
+        }
+
+        if(isRoot) {
             lineColor = 'rgba(0,0,0,0)';
             lineWidth = 0;
         } else {
             lineColor = Lib.castOption(trace, ptNumber, 'marker.line.color') || Color.defaultLine;
             lineWidth = Lib.castOption(trace, ptNumber, 'marker.line.width') || 0;
         }
-
-        opacity =
-            trace._hasColorscale || helpers.isLeaf(pt) ? 1 :
-            pt.onPathbar ? trace.pathbar.opacity :
-                depthFade(pt.data.depth - trace._entryDepth);
     }
 
     s.style('stroke-width', lineWidth)
-        .call(Color.fill, cdi.color)
+        .call(Color.fill, fillColor)
         .call(Color.stroke, lineColor)
-        .style('opacity', opacity);
+        .style('opacity', null);
```

Comment:
Do we need to unset `opacity` here? That is, do we set `s.style('opacity', /**/)` elsewhere?

Reply:
Cool. Let's :hocho: it.

Reply:
Dropped in https://github.com/plotly/plotly.js/pull/4242/commits/3b104b143050a6bbeb0853486dd2ac5a3a82802c but then the opacity 0 is actually used for handling root without color special case in https://github.com/plotly/plotly.js/pull/4242/commits/1dff468da972eb9417c8b3c690e9accfafdf30eb.